# Deep Learning Computing Lab
## DSBA 6190

---

Name: Muralidharan Singaravel

Student ID: 801 059 720

| Total Points Earned | Total Points Possible |
|---------------------|-----------------------|
|                     | 100                   |

## Section 1a) Prep your Deep Learning Framework

| Points Earned | Points Possible |
|---------------|-----------------|
|               | 5               |

1. Pick a deep learning framework to use in Databricks/Spark
  - Choices:
    - [Keras](https://docs.azuredatabricks.net/applications/deep-learning/single-node-training/keras.html)  ← Recommended for this Assignment
    - [Tensorflow](https://docs.azuredatabricks.net/applications/deep-learning/single-node-training/tensorflow.html)
    - [CNTK](https://docs.azuredatabricks.net/applications/deep-learning/single-node-training/cntk.html)
    - [MXNET](https://docs.azuredatabricks.net/applications/deep-learning/single-node-training/mxnet.html)
    - [PyTorch](https://docs.azuredatabricks.net/applications/deep-learning/single-node-training/pytorch.html)
2. Load in the required libraries for your selected framework (below)

Which framework did you choose?: Keras

In [3]:

from __future__ import print_function
import keras

from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten
from keras.layers import Conv2D, MaxPooling2D
from keras import backend as K

from keras.datasets import fashion_mnist



## Section 1b) Pick your Dataset and Read it in

| Points Earned | Points Possible |
|---------------|-----------------|
|               | 10              |

1. Pick a dataset to use for building a Neural Network model.

_Note: It is likely beneficial to pick a dataset that's included/documented with your deep learning framework so that you can use example code to get started._

_Note: If you pick a dataset that is meant for your selected framework, you may be able to call it from the library itself or you may have to download it and put it in Blob storage._

Resources:
  - [Keras Datasets](https://keras.io/datasets/)
  - [Tensorflow Datasets](https://github.com/tensorflow/datasets/blob/master/docs/datasets.md)
  - [CNTK Datasets](https://github.com/Microsoft/CNTK/tree/master/Examples/Image/DataSets)

Description of your Data: Fashion-MNIST

Data Source: Keras


Fashion-MNIST database of fashion articles
Dataset of 60,000 28x28 grayscale images of 10 fashion categories, along with a test set of 10,000 images.

## Section 2) Shape your Data

| Points Earned | Points Possible |
|---------------|-----------------|
|               | 30              |

1. Transform and prepare your data for use in neural network creation.

In [6]:
# the data, split between train and test sets
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

Downloading data from http://fashion-mnist.s3-website.eu-central-1.amazonaws.com/train-labels-idx1-ubyte.gz

 8192/29515 [=======>......................] - ETA: 0s
16384/29515 [===============>..............] - ETA: 0s
32768/29515 [=================================] - 0s 3us/step
Downloading data from http://fashion-mnist.s3-website.eu-central-1.amazonaws.com/train-images-idx3-ubyte.gz

 8192/26421880 [..............................] - ETA: 0s
 16384/26421880 [..............................] - ETA: 2:27
 49152/26421880 [..............................] - ETA: 1:38
 106496/26421880 [..............................] - ETA: 1:08
 229376/26421880 [..............................] - ETA: 41s 
 466944/26421880 [..............................] - ETA: 25s
 950272/26421880 [>.............................] - ETA: 14s
 1908736/26421880 [=>............................] - ETA: 7s 
 3825664/26421880 [===>..........................] - ETA: 4s
 6938624/26421880 [======>.......................] - ETA: 2s
10059776/26421880 [==========>...................] - ETA: 1s
13148160/26421880 [=============>................] - ETA: 0s
13230080/26421880 [==============>...............] - ETA: 1s
16351232/26421880 [=================>............] - ETA: 0s
18014208/26421880 [===================>..........] - ETA: 0s
19505152/26421880 [=====================>........] - ETA: 0s
22552576/26421880 [========================>.....] - ETA: 0s
24190976/26421880 [==========================>...] - ETA: 0s
25788416/26421880 [============================>.] - ETA: 0s
26427392/26421880 [==============================] - 1s 0us/step
Downloading data from http://fashion-mnist.s3-website.eu-central-1.amazonaws.com/t10k-labels-idx1-ubyte.gz

8192/5148 [===============================================] - 0s 0us/step
Downloading data from http://fashion-mnist.s3-website.eu-central-1.amazonaws.com/t10k-images-idx3-ubyte.gz

 8192/4422102 [..............................] - ETA: 0s
 16384/4422102 [..............................] - ETA: 24s
 49152/4422102 [..............................] - ETA: 16s
 106496/4422102 [..............................] - ETA: 11s
 229376/4422102 [>.............................] - ETA: 6s 
 466944/4422102 [==>...........................] - ETA: 3s
 950272/4422102 [=====>........................] - ETA: 1s
1908736/4422102 [===========>..................] - ETA: 0s
3833856/4422102 [=========================>....] - ETA: 0s
4423680/4422102 [==============================] - 1s 0us/step

In [7]:
# input image dimensions
img_rows, img_cols = 28, 28
num_classes = 10

if K.image_data_format() == 'channels_first':
    x_train = x_train.reshape(x_train.shape[0], 1, img_rows, img_cols)
    x_test = x_test.reshape(x_test.shape[0], 1, img_rows, img_cols)
    input_shape = (1, img_rows, img_cols)
else:
    x_train = x_train.reshape(x_train.shape[0], img_rows, img_cols, 1)
    x_test = x_test.reshape(x_test.shape[0], img_rows, img_cols, 1)
    input_shape = (img_rows, img_cols, 1)

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')
x_train /= 255
x_test /= 255
print('x_train shape:', x_train.shape)
print(x_train.shape[0], 'train samples')
print(x_test.shape[0], 'test samples')


# convert class vectors to binary class matrices
y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)




x_train shape: (60000, 28, 28, 1)
60000 train samples
10000 test samples

## Section 3a) Create a Neural Network Model

| Points Earned | Points Possible |
|---------------|-----------------|
|               | 40              |

1. Create at least 1 neural network using your framework and dataset.

2. Explain the type and architecture of your model below. _(For example: "I'm using a CNN with 3 hidden layers.")_

Description of the your neural network: I'm using a CNN with 8 hidden layers

_Note: You can also provide a diagram if you want. You'll need to use some sort of tool such as [Tensorflow Playground](https://playground.tensorflow.org/) or draw it in some diagramming software._

In [9]:


model = Sequential()
model.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Dropout(0.25))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.5))
model.add(Dense(num_classes, activation='softmax'))

model.compile(loss=keras.losses.categorical_crossentropy,
              optimizer=keras.optimizers.Adadelta(),
              metrics=['accuracy'])


WARNING:tensorflow:From /databricks/python/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:74: The name tf.get_default_graph is deprecated. Please use tf.compat.v1.get_default_graph instead.

WARNING:tensorflow:From /databricks/python/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:517: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.

WARNING:tensorflow:From /databricks/python/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:4138: The name tf.random_uniform is deprecated. Please use tf.random.uniform instead.

/databricks/python/lib/python3.7/site-packages/tensorflow/python/ops/nn_ops.py:61: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
 elif not isinstance(value, collections.Sized):
WARNING:tensorflow:From /databricks/python/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:3976: The name tf.nn.max_pool is deprecated. Please use tf.nn.max_pool2d instead.

WARNING:tensorflow:From /databricks/python/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:133: The name tf.placeholder_with_default is deprecated. Please use tf.compat.v1.placeholder_with_default instead.

WARNING:tensorflow:From /databricks/python/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:3445: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.
Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.
WARNING:tensorflow:From /databricks/python/lib/python3.7/site-packages/keras/optimizers.py:790: The name tf.train.Optimizer is deprecated. Please use tf.compat.v1.train.Optimizer instead.

WARNING:tensorflow:From /databricks/python/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py:3295: The name tf.log is deprecated. Please use tf.math.log instead.

In [10]:
# Summary
model.summary()

_________________________________________________________________
Layer (type) Output Shape Param # 
=================================================================
conv2d_1 (Conv2D) (None, 26, 26, 32) 320 
_________________________________________________________________
conv2d_2 (Conv2D) (None, 24, 24, 64) 18496 
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 12, 12, 64) 0 
_________________________________________________________________
dropout_1 (Dropout) (None, 12, 12, 64) 0 
_________________________________________________________________
flatten_1 (Flatten) (None, 9216) 0 
_________________________________________________________________
dense_1 (Dense) (None, 128) 1179776 
_________________________________________________________________
dropout_2 (Dropout) (None, 128) 0 
_________________________________________________________________
dense_2 (Dense) (None, 10) 1290 
=================================================================
Total params: 1,199,882
Trainable params: 1,199,882
Non-trainable params: 0
_________________________________________________________________

In [11]:
#number of layers
model.layers

Out[9]: [<keras.layers.convolutional.Conv2D at 0x7f0e4966d9b0>,
 <keras.layers.convolutional.Conv2D at 0x7f0e9f7acfd0>,
 <keras.layers.pooling.MaxPooling2D at 0x7f0e9f7ac320>,
 <keras.layers.core.Dropout at 0x7f0e46ac2c50>,
 <keras.layers.core.Flatten at 0x7f0e9f7acf98>,
 <keras.layers.core.Dense at 0x7f0e46ac2c88>,
 <keras.layers.core.Dropout at 0x7f0e46a7bc88>,
 <keras.layers.core.Dense at 0x7f0e469ce5c0>]

In [12]:
len(model.layers)

Out[10]: 8

In [13]:
batch_size = 128

epochs = 4
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))


WARNING:tensorflow:From /databricks/python/lib/python3.7/site-packages/tensorflow/python/ops/math_grad.py:1250: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where
Train on 60000 samples, validate on 10000 samples
Epoch 1/4

 128/60000 [..............................] - ETA: 15:55 - loss: 2.2859 - acc: 0.1250
 512/60000 [..............................] - ETA: 4:04 - loss: 2.1143 - acc: 0.2539 
 896/60000 [..............................] - ETA: 2:22 - loss: 2.0020 - acc: 0.2935
 1280/60000 [..............................] - ETA: 1:41 - loss: 1.8580 - acc: 0.3406
 1664/60000 [..............................] - ETA: 1:19 - loss: 1.7745 - acc: 0.3750
 2176/60000 [>.............................] - ETA: 1:01 - loss: 1.6469 - acc: 0.4237
 2560/60000 [>.............................] - ETA: 53s - loss: 1.5600 - acc: 0.4551 
 2944/60000 [>.............................] - ETA: 47s - loss: 1.4928 - acc: 0.4800
 3456/60000 [>.............................] - ETA: 40s - loss: 1.4092 - acc: 0.5061
 3968/60000 [>.............................] - ETA: 36s - loss: 1.3394 - acc: 0.5280
 4480/60000 [=>............................] - ETA: 32s - loss: 1.2827 - acc: 0.5451
 4992/60000 [=>............................] - ETA: 29s - loss: 1.2342 - acc: 0.5611
 5504/60000 [=>............................] - ETA: 27s - loss: 1.1871 - acc: 0.5803
 6016/60000 [==>...........................] - ETA: 25s - loss: 1.1460 - acc: 0.5928
 6528/60000 [==>...........................] - ETA: 23s - loss: 1.1102 - acc: 0.6059
 7040/60000 [==>...........................] - ETA: 21s - loss: 1.0826 - acc: 0.6168
 7552/60000 [==>...........................] - ETA: 20s - loss: 1.0589 - acc: 0.6247
 8064/60000 [===>..........................] - ETA: 19s - loss: 1.0332 - acc: 0.6334
 8576/60000 [===>..........................] - ETA: 18s - loss: 1.0144 - acc: 0.6398
 9088/60000 [===>..........................] - ETA: 17s - loss: 0.9937 - acc: 0.6472
 9600/60000 [===>..........................] - ETA: 16s - loss: 0.9746 - acc: 0.6532
10112/60000 [====>.........................] - ETA: 16s - loss: 0.9625 - acc: 0.6573
10624/60000 [====>.........................] - ETA: 15s - loss: 0.9463 - acc: 0.6635
11136/60000 [====>.........................] - ETA: 14s - loss: 0.9333 - acc: 0.6677
11648/60000 [====>.........................] - ETA: 14s - loss: 0.9211 - acc: 0.6720
12160/60000 [=====>........................] - ETA: 13s - loss: 0.9071 - acc: 0.6772
12672/60000 [=====>........................] - ETA: 13s - loss: 0.8950 - acc: 0.6815
13184/60000 [=====>........................] - ETA: 12s - loss: 0.8863 - acc: 0.6848
13696/60000 [=====>........................] - ETA: 12s - loss: 0.8726 - acc: 0.6894
14208/60000 [======>.......................] - ETA: 12s - loss: 0.8600 - acc: 0.6934
14720/60000 [======>.......................] - ETA: 11s - loss: 0.8494 - acc: 0.6969
15232/60000 [======>.......................] - ETA: 11s - loss: 0.8410 - acc: 0.7000
15744/60000 [======>.......................] - ETA: 11s - loss: 0.8330 - acc: 0.7031
16256/60000 [=======>......................] - ETA: 10s - loss: 0.8269 - acc: 0.7049
16768/60000 [=======>......................] - ETA: 10s - loss: 0.8194 - acc: 0.7083
17280/60000 [=======>......................] - ETA: 10s - loss: 0.8120 - acc: 0.7112
17792/60000 [=======>......................] - ETA: 9s - loss: 0.8054 - acc: 0.7132 
18304/60000 [========>.....................] - ETA: 9s - loss: 0.7987 - acc: 0.7157
18816/60000 [========>.....................] - ETA: 9s - loss: 0.7908 - acc: 0.7185
19328/60000 [========>.....................] - ETA: 9s - loss: 0.7842 - acc: 0.7213
19840/60000 [========>.....................] - ETA: 8s - loss: 0.7774 - acc: 0.7238
20352/60000 [=========>....................] - ETA: 8s - loss: 0.7699 - acc: 0.7266
20864/60000 [=========>.......

In [14]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss (epochs = 4):', score[0])
print('Test accuracy(epochs = 4):', score[1])

Test loss (epochs = 4): 0.2560270010471344
Test accuracy(epochs = 4): 0.9061

In [15]:
#increased the epochs to 16 - The number of epochs is the number of complete passes through the training dataset. 
batch_size = 128

epochs = 16
model.fit(x_train, y_train,
          batch_size=batch_size,
          epochs=epochs,
          verbose=1,
          validation_data=(x_test, y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/16

 128/60000 [..............................] - ETA: 7s - loss: 0.2546 - acc: 0.8984
 640/60000 [..............................] - ETA: 7s - loss: 0.2370 - acc: 0.9031
 1152/60000 [..............................] - ETA: 6s - loss: 0.2376 - acc: 0.9062
 1664/60000 [..............................] - ETA: 6s - loss: 0.2361 - acc: 0.9081
 2176/60000 [>.............................] - ETA: 6s - loss: 0.2378 - acc: 0.9076
 2688/60000 [>.............................] - ETA: 6s - loss: 0.2403 - acc: 0.9074
 3200/60000 [>.............................] - ETA: 6s - loss: 0.2502 - acc: 0.9044
 3712/60000 [>.............................] - ETA: 6s - loss: 0.2473 - acc: 0.9044
 4224/60000 [=>............................] - ETA: 6s - loss: 0.2540 - acc: 0.8999
 4736/60000 [=>............................] - ETA: 6s - loss: 0.2526 - acc: 0.9014
 5248/60000 [=>............................] - ETA: 6s - loss: 0.2505 - acc: 0.9022
 5760/60000 [=>............................] - ETA: 6s - loss: 0.2533 - acc: 0.9017
 6272/60000 [==>...........................] - ETA: 6s - loss: 0.2514 - acc: 0.9029
 6784/60000 [==>...........................] - ETA: 6s - loss: 0.2507 - acc: 0.9027
 7296/60000 [==>...........................] - ETA: 6s - loss: 0.2509 - acc: 0.9032
 7808/60000 [==>...........................] - ETA: 6s - loss: 0.2513 - acc: 0.9042
 8320/60000 [===>..........................] - ETA: 6s - loss: 0.2488 - acc: 0.9049
 8832/60000 [===>..........................] - ETA: 5s - loss: 0.2490 - acc: 0.9047
 9344/60000 [===>..........................] - ETA: 5s - loss: 0.2510 - acc: 0.9048
 9856/60000 [===>..........................] - ETA: 5s - loss: 0.2491 - acc: 0.9058
10368/60000 [====>.........................] - ETA: 5s - loss: 0.2482 - acc: 0.9063
10880/60000 [====>.........................] - ETA: 5s - loss: 0.2506 - acc: 0.9053
11392/60000 [====>.........................] - ETA: 5s - loss: 0.2475 - acc: 0.9066
11904/60000 [====>.........................] - ETA: 5s - loss: 0.2475 - acc: 0.9070
12416/60000 [=====>........................] - ETA: 5s - loss: 0.2480 - acc: 0.9066
12928/60000 [=====>........................] - ETA: 5s - loss: 0.2483 - acc: 0.9063
13440/60000 [=====>........................] - ETA: 5s - loss: 0.2475 - acc: 0.9064
13952/60000 [=====>........................] - ETA: 5s - loss: 0.2478 - acc: 0.9064
14464/60000 [======>.......................] - ETA: 5s - loss: 0.2490 - acc: 0.9060
14976/60000 [======>.......................] - ETA: 5s - loss: 0.2492 - acc: 0.9058
15488/60000 [======>.......................] - ETA: 5s - loss: 0.2488 - acc: 0.9059
16000/60000 [=======>......................] - ETA: 5s - loss: 0.2504 - acc: 0.9048
16512/60000 [=======>......................] - ETA: 5s - loss: 0.2503 - acc: 0.9049
17024/60000 [=======>......................] - ETA: 5s - loss: 0.2500 - acc: 0.9051
17536/60000 [=======>......................] - ETA: 4s - loss: 0.2497 - acc: 0.9051
18048/60000 [========>.....................] - ETA: 4s - loss: 0.2517 - acc: 0.9047
18560/60000 [========>.....................] - ETA: 4s - loss: 0.2514 - acc: 0.9051
19072/60000 [========>.....................] - ETA: 4s - loss: 0.2508 - acc: 0.9057
19584/60000 [========>.....................] - ETA: 4s - loss: 0.2504 - acc: 0.9061
20096/60000 [=========>....................] - ETA: 4s - loss: 0.2492 - acc: 0.9065
20608/60000 [=========>....................] - ETA: 4s - loss: 0.2494 - acc: 0.9067
21120/60000 [=========>....................] - ETA: 4s - loss: 0.2494 - acc: 0.9069
21632/60000 [=========>....................] - ETA: 4s - loss: 0.2510 - acc: 0.9063
22144/60000 [==========>...................] - ETA: 4s - loss: 0.2511 - acc: 0.9062
22656/60000 [==========>...................] - ETA: 4s - loss: 0.2504 - acc: 0.9067
23168/60000 [==========>...................] - ETA: 4s - loss: 0.2515 - acc: 0.9064
23680/60000 [==========>...................] - ETA: 4s - loss: 0.2519 - acc:

In [16]:
score = model.evaluate(x_test, y_test, verbose=0)
print('Test loss (epochs = 16):', score[0])
print('Test accuracy(epochs = 16):', score[1])

Test loss (epochs = 16): 0.22832045611143112
Test accuracy(epochs = 16): 0.9231

## Section 4) Evaluate your Model

| Points Earned | Points Possible |
|---------------|-----------------|
|               | 15              |

1. Evaluate and discuss your model.
I ran the model with 4 and 16 epochs with the batch size of 128.

2. How is it performing?
The model with 4 epochs has accuracy of 0.90 and took 32 seconds to complete where as the model with 16 epochs had accuracy of 0.93 and took 2 minutes to complete but increasing the epochs will result in overfitting.

3. What things could you do to improve its performance?

Few ways to improve the performace are:

a. Rescale and transform the  Data by converting data to right format and Pre-process the data according to the needs.

b. Do a feature selection, if it can be done on the dataset

c. Re-evaluate the varous neural network methods like LVQ, MLP, CNN, LSTM, hybrids, etc.,  is it really the best technique one could have chosen 

d. Choose a different sampling methods like -  k-fold cross validation.